python/lsst/pipe/base/makeLightWeightButler.py

```diff
@@ -0,0 +1,246 @@
+    # lightweight butler
+    exports: Set[DatasetRef] = set()
+
+    # inserts is the mapping of DatasetType to dataIds for what is to be
+    # inserted into the registry. These are the products that are expected
+    # to be produced during processing of the QuantumGraph
+    inserts: DefaultDict[DatasetType, Set[DataCoordinate]] = defaultdict(set)
+
+    n: QuantumNode
+    for quantum in (n.quantum for n in graph):
+        for attrName in ("initInputs", "inputs", "outputs"):
+            attr: Mapping[DatasetType, Union[DatasetRef, List[DatasetRef]]] = getattr(quantum, attrName)
+
+            for type, refs in attr.items():
+                # This if block is because init inputs has a different
+                # signature for its items
+                if not isinstance(refs, list):
+                    refs = [refs]
+                # iterate over all the references, if it has an id, it
+                # means it exists and should be exported, if not it should
+                # be inserted into the new registry
+                for ref in refs:
+                    if ref.isComponent():
+                        # We can't insert a component, and a component will
+                        # be part of some other upstream dataset, so it
+                        # should be safe to skip them here
+                        continue
+
+                    if ref.id is not None:
+                        exports.add(ref)
+                    else:
+                        inserts[type].add(ref.dataId)
+    return exports, inserts
+
+
+def _discoverCollections(butler: Butler, collections: Iterable[str]) -> set[str]:
+    # Recurse through any discovered collections to make sure all collections
+    # are exported. This exists because I ran into a situation where some
+    # collections were not properly being discovered and exported. This
+    # method may be able to be removed in the future if collection export
+    # logic changes
+    collections = set(collections)
+    while True:
+        discoveredCollections = set(butler.registry.queryCollections(collections, flattenChains=True,
+                                                                     includeChains=True))
+        if discoveredCollections - collections:
+            collections = discoveredCollections
+        else:
+            break
+    return collections
+
+
+def _export(butler: Butler, collections: Optional[Iterable[str]], exports: Set[DatasetRef]) -> io.StringIO:
+    # This exports the datasets that exist in the input butler using
+    # daf butler objects, however it reaches in deep and does not use the
+    # public methods so that it can export it to a string buffer and skip
+    # disk access.
+    yamlBuffer = io.StringIO()
+    # Yaml is hard coded, since the class controls both ends of the
+    # export/import
+    BackendClass = getClassOf(butler._config["repo_transfer_formats"]["yaml"]["export"])
```

Comment:
I think this can be simplified or made more efficient as `_config["repo_transfer_formats", "yaml", "export"]`